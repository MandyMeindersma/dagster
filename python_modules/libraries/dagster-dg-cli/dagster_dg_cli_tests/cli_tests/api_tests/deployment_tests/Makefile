.PHONY: rerecord
rerecord:
	dagster-dev dg-api-record deployment
