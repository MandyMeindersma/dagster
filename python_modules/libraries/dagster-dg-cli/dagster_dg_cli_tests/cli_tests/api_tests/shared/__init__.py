"""Shared utilities for API test fixtures."""
