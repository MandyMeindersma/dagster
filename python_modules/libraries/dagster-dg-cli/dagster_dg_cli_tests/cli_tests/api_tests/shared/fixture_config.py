"""Configuration for API fixture recording commands."""

from dataclasses import dataclass
from typing import Optional


@dataclass
class FixtureCommand:
    """Configuration for a single fixture recording command."""

    command: str
    depends_on: Optional[list[str]] = None  # Dependencies (run IDs, etc.)
