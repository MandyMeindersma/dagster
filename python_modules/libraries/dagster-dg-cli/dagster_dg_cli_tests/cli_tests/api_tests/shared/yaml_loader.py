"""YAML loading utilities for fixture command configurations."""

from pathlib import Path

import yaml

from dagster_dg_cli_tests.cli_tests.api_tests.shared.fixture_config import FixtureCommand


def load_fixture_commands_from_yaml(yaml_file: Path) -> dict[str, FixtureCommand]:
    """Load fixture commands from a YAML file.

    Args:
        yaml_file: Path to YAML file containing fixture command definitions

    Returns:
        Dictionary mapping fixture names to FixtureCommand objects

    Raises:
        ValueError: If YAML file doesn't exist or has invalid format
    """
    if not yaml_file.exists():
        raise ValueError(f"YAML fixture file not found: {yaml_file}")

    try:
        with open(yaml_file) as f:
            yaml_data = yaml.safe_load(f) or {}
    except yaml.YAMLError as e:
        raise ValueError(f"Invalid YAML in {yaml_file}: {e}")

    fixture_commands = {}
    for fixture_name, config_data in yaml_data.items():
        if not isinstance(config_data, dict):
            raise ValueError(
                f"Invalid fixture config for '{fixture_name}': expected dict, got {type(config_data)}"
            )

        if "command" not in config_data:
            raise ValueError(f"Missing 'command' field for fixture '{fixture_name}'")

        fixture_commands[fixture_name] = FixtureCommand(
            command=config_data["command"], depends_on=config_data.get("depends_on")
        )

    return fixture_commands
