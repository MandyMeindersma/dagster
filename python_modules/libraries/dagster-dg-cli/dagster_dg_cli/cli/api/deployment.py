"""Deployment API commands following GitHub CLI patterns."""

import sys

import click
from dagster_dg_core.utils import DgClickCommand, DgClickGroup
from dagster_dg_core.utils.telemetry import cli_telemetry_wrapper

from dagster_dg_cli.cli.api.formatters import format_deployments
from dagster_dg_cli.cli.api.shared import format_error_for_output, get_config_or_error
from dagster_dg_cli.dagster_plus_api.api.deployments import DgApiDeploymentApi


@click.command(name="list", cls=DgClickCommand, unlaunched=True)
@click.option(
    "--json",
    "output_json",
    is_flag=True,
    help="Output in JSON format for machine readability",
)
@cli_telemetry_wrapper
def list_deployments_command(output_json: bool) -> None:
    """List all deployments in the organization."""
    config = get_config_or_error()
    api = DgApiDeploymentApi(config)

    try:
        deployments = api.list_deployments()
        output = format_deployments(deployments, as_json=output_json)
        click.echo(output)
    except Exception as e:
        error_output, exit_code = format_error_for_output(e, output_json)
        click.echo(error_output, err=True)
        sys.exit(exit_code)


@click.group(
    name="deployment",
    cls=DgClickGroup,
    unlaunched=True,
    commands={
        "list": list_deployments_command,
    },
)
def deployment_group():
    """Manage deployments in Dagster Plus."""
