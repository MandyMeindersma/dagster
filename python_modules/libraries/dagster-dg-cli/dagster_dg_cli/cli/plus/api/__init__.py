"""API commands for interacting with Dagster Plus REST-like interface."""

import click
from dagster_dg_core.utils import DgClickGroup

from dagster_dg_cli.cli.plus.api.deployments import deployments_group


@click.group(
    name="api",
    cls=DgClickGroup,
    unlaunched=True,
    commands={
        "deployments": deployments_group,
    },
)
def api_group():
    """Make REST-like API calls to Dagster Plus."""
