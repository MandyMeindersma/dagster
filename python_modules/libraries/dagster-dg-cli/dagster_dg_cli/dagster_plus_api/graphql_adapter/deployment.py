"""GraphQL implementation for deployment operations."""

from typing import TYPE_CHECKING, Any, Optional

from dagster_shared.plus.config import DagsterPlusCliConfig

from dagster_dg_cli.utils.plus.gql_client import DagsterPlusGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.dagster_plus_api.schemas.deployment import DeploymentList

# GraphQL queries
LIST_DEPLOYMENTS_QUERY = """
query ListDeployments {
    fullDeployments {
        deploymentName
        deploymentId
        deploymentType
    }
}
"""


def process_deployments_response(
    graphql_response: dict[str, Any], limit: Optional[int] = None
) -> "DeploymentList":
    """Process GraphQL response into DeploymentList.

    This is a pure function that can be easily tested without mocking GraphQL clients.

    Args:
        graphql_response: Raw GraphQL response containing "fullDeployments"
        limit: Optional limit to apply to results

    Returns:
        DeploymentList: Processed deployment data
    """
    # Import pydantic models only when needed
    from dagster_dg_cli.dagster_plus_api.schemas.deployment import (
        Deployment,
        DeploymentList,
        DeploymentType,
    )

    deployments_data = graphql_response.get("fullDeployments", [])

    deployments = [
        Deployment(
            id=d["deploymentId"],
            name=d["deploymentName"],
            type=DeploymentType[d["deploymentType"]],
        )
        for d in deployments_data
    ]

    # Apply limit if specified - GraphQL fullDeployments does not support limit parameter
    # so we must filter client-side
    if limit:
        deployments = deployments[:limit]

    return DeploymentList(
        items=deployments,
        total=len(deployments),
    )


def list_deployments_via_graphql(
    config: DagsterPlusCliConfig,
    limit: Optional[int] = None,
) -> "DeploymentList":
    """Fetch deployments using GraphQL.
    This is an implementation detail that can be replaced with REST calls later.
    """
    client = DagsterPlusGraphQLClient.from_config(config)
    result = client.execute(LIST_DEPLOYMENTS_QUERY)

    return process_deployments_response(result, limit=limit)
