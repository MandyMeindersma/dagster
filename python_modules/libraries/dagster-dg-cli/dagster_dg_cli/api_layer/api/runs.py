"""Run endpoints - REST-like interface."""

from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional

from dagster_dg_cli.api_layer.graphql_adapter.run import get_run_via_graphql, list_runs_via_graphql
from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.run import Run, RunList


@dataclass(frozen=True)
class DgApiRunApi:
    client: IGraphQLClient

    def list_runs(
        self,
        limit: Optional[int] = None,
        status: Optional[str] = None,
        job_name: Optional[str] = None,
        cursor: Optional[str] = None,
    ) -> "RunList":
        """List runs with optional filtering.

        Args:
            limit: Maximum number of runs to return
            status: Filter by run status
            job_name: Filter by job name
            cursor: Pagination cursor

        Returns:
            RunList: List of runs matching the filters
        """
        return list_runs_via_graphql(
            client=self.client,
            limit=limit,
            status=status,
            job_name=job_name,
            cursor=cursor,
        )

    def get_run(self, run_id: str) -> "Run":
        """Get a specific run by ID.

        Args:
            run_id: The run ID to fetch

        Returns:
            Run: The requested run

        Raises:
            Exception: If run not found or API error occurs
        """
        return get_run_via_graphql(client=self.client, run_id=run_id)
