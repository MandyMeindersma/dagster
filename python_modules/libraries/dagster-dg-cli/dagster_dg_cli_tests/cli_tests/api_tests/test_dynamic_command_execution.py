"""Dynamic test execution from scenarios.yaml fixtures.

This test module automatically discovers all scenarios defined in scenarios.yaml files
across API test domains and runs them against their recorded GraphQL response fixtures.
"""

import json
from collections.abc import Iterator
from pathlib import Path
from typing import Any
from unittest.mock import patch

import pytest
from click.testing import CliRunner

from dagster_dg_cli_tests.cli_tests.api_tests.shared.yaml_loader import (
    load_fixture_scenarios_from_yaml,
)


def discover_scenario_fixtures() -> Iterator[tuple[str, str, str]]:
    """Discover all scenario fixtures across API test domains.

    Returns:
        Iterator of (domain, fixture_name, command) tuples
    """
    api_tests_dir = Path(__file__).parent

    # Find all domain test directories
    domain_dirs = [d for d in api_tests_dir.iterdir() if d.is_dir() and d.name.endswith("_tests")]

    for domain_dir in domain_dirs:
        scenarios_yaml = domain_dir / "fixtures" / "scenarios.yaml"
        if not scenarios_yaml.exists():
            continue

        domain = domain_dir.name.replace("_tests", "")
        fixture_scenarios = load_fixture_scenarios_from_yaml(scenarios_yaml)

        for fixture_name, fixture_config in fixture_scenarios.items():
            yield (domain, fixture_name, fixture_config.command)


def get_command_mapping() -> dict[str, Any]:
    """Map command strings to their corresponding Click command objects.

    Returns:
        Dictionary mapping command patterns to Click commands
    """
    # Import CLI commands
    from dagster_dg_cli.cli.api.deployment import list_deployments_command

    return {
        "dg api deployment list": list_deployments_command,
        # TODO: Add more command mappings as needed
    }


def parse_command_string(command: str) -> tuple[Any, list[str]]:
    """Parse a command string to extract the Click command and arguments.

    Args:
        command: Full command string (e.g., "dg api deployment list --json")

    Returns:
        Tuple of (click_command, args_list)

    Raises:
        ValueError: If command is not recognized
    """
    command_mapping = get_command_mapping()

    # Split command and look for matches
    parts = command.split()

    # Try to find the longest matching command prefix
    for i in range(len(parts), 0, -1):
        prefix = " ".join(parts[:i])
        if prefix in command_mapping:
            click_command = command_mapping[prefix]
            remaining_args = parts[i:]
            return click_command, remaining_args

    raise ValueError(f"No matching Click command found for: {command}")


def load_fixture_graphql_responses(domain: str, fixture_name: str) -> list[dict[str, Any]]:
    """Load GraphQL response fixtures for a given domain and fixture.

    Args:
        domain: API domain (e.g., 'deployment', 'asset')
        fixture_name: Name of the fixture scenario

    Returns:
        List of GraphQL response dictionaries
    """
    if domain == "deployment":
        from dagster_dg_cli_tests.cli_tests.api_tests.deployment_tests.fixtures import (
            load_deployment_response_fixture,
        )

        return load_deployment_response_fixture(fixture_name)
    # TODO: Add other domains as they become available
    else:
        raise ValueError(f"Unknown domain: {domain}")


class TestDynamicCommandExecution:
    """Test all commands from YAML fixtures against recorded GraphQL responses."""

    @pytest.mark.parametrize("domain,fixture_name,command", list(discover_scenario_fixtures()))
    def test_command_execution(self, domain: str, fixture_name: str, command: str, snapshot):
        """Test executing a command against its recorded GraphQL responses."""
        try:
            # Parse the command
            click_command, args = parse_command_string(command)

            # Load GraphQL responses for this fixture
            graphql_responses = load_fixture_graphql_responses(domain, fixture_name)

            # Mock the GraphQL client based on domain
            if domain == "deployment":
                result = self._test_deployment_command(
                    click_command, args, graphql_responses, fixture_name
                )

                # Create a unique snapshot name based on fixture and output type
                output_type = "json" if "--json" in args else "text"
                snapshot_name = f"{domain}_{fixture_name}_{output_type}"

                # Use syrupy to snapshot the CLI output with custom name
                if "--json" in args:
                    # For JSON output, parse and snapshot the structure
                    try:
                        parsed_output = json.loads(result.output)
                        assert parsed_output == snapshot(name=snapshot_name)
                    except json.JSONDecodeError:
                        # For error cases, snapshot the raw output
                        assert result.output == snapshot(name=snapshot_name)
                else:
                    # For text output, snapshot the raw CLI output
                    assert result.output == snapshot(name=snapshot_name)

                # Keep existing exit code assertions for safety
                if "error" not in fixture_name.lower():
                    assert result.exit_code == 0
            else:
                pytest.skip(f"Domain '{domain}' not yet supported in dynamic tests")

        except Exception as e:
            pytest.fail(f"Failed to execute command '{command}' for fixture '{fixture_name}': {e}")

    def _test_deployment_command(
        self, click_command: Any, args: list[str], graphql_responses: list[dict], fixture_name: str
    ):
        """Test a deployment domain command with mocked GraphQL responses."""
        with patch("dagster_dg_cli.cli.api.shared.get_config_or_error") as mock_config:
            with patch(
                "dagster_dg_cli.dagster_plus_api.graphql_adapter.deployment.DagsterPlusGraphQLClient.from_config"
            ) as mock_client_class:
                # Setup mocks
                mock_config.return_value = "mock-config"
                mock_client = mock_client_class.return_value
                mock_client.execute.return_value = graphql_responses[0]

                # Execute command
                runner = CliRunner()
                result = runner.invoke(click_command, args)

                # Return the result for snapshot testing
                return result
