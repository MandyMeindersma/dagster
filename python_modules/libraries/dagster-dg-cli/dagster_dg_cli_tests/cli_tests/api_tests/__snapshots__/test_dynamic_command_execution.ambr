# serializer version: 1
# name: TestDynamicCommandExecution.test_command_execution[deployment_empty_deployments_json]
  dict({
    'items': list([
    ]),
    'total': 0,
  })
# ---
# name: TestDynamicCommandExecution.test_command_execution[deployment_success_multiple_deployments_json]
  dict({
    'items': list([
      dict({
        'id': 6935,
        'name': 'uat',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 6936,
        'name': 'excessive',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 84915,
        'name': 'ben-test',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 37,
        'name': 'ecs',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 6934,
        'name': 'staging',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 80,
        'name': 'prod',
        'type': 'PRODUCTION',
      }),
    ]),
    'total': 6,
  })
# ---
# name: TestDynamicCommandExecution.test_command_execution[deployment_success_multiple_deployments_text_text]
  '''
  Name: uat
  ID: 6935
  Type: PRODUCTION
  
  Name: excessive
  ID: 6936
  Type: PRODUCTION
  
  Name: ben-test
  ID: 84915
  Type: PRODUCTION
  
  Name: ecs
  ID: 37
  Type: PRODUCTION
  
  Name: staging
  ID: 6934
  Type: PRODUCTION
  
  Name: prod
  ID: 80
  Type: PRODUCTION
  
  '''
# ---
